### PR TITLE
feat: split rpcContext

### PR DESCRIPTION
**Pull request by aamingaa**
This is the pure version

### CODE REVIEW

**Review by AlbumenJ — CHANGES_REQUESTED** [review]
What is the difference between #10793

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/filter/support/ConsumerContextFilter.java [hunk 1]
```diff
@@ -116,8 +114,9 @@ public Result invoke(Invoker<?> invoker, Invocation invocation) throws RpcExcept
     @Override
     public void onResponse(Result appResponse, Invoker<?> invoker, Invocation invocation) {
         // pass attachments to result
-        RpcContext.getServerContext().setObjectAttachments(appResponse.getObjectAttachments());
-
+        Map<String, Object> map = appResponse.getObjectAttachments();
+        RpcContext.getClientResponseContext().setObjectAttachments(map);
+        RpcContext.getServerContext().setObjectAttachments(map);
```
Comment (AlbumenJ) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/filter/support/ConsumerContextFilter.java:119]: should not call this method here

File: dubbo-compatible/src/main/java/com/alibaba/dubbo/rpc/RpcContext.java [hunk 2]
```diff
@@ -47,9 +47,19 @@ public static RpcContext getContext() {
...
+
+    public static RpcContext getServerResponseContext() {
+        return new RpcContext(org.apache.dubbo.rpc.RpcContext.getServerResponseContext());
+    }
+    public static void removeClientResponseContext() {
+        org.apache.dubbo.rpc.RpcContext.removeClientResponseContext();
+    }
 
-    public static void removeServerContext() {
-        org.apache.dubbo.rpc.RpcContext.removeServerContext();
```
Comment (AlbumenJ) [inline dubbo-compatible/src/main/java/com/alibaba/dubbo/rpc/RpcContext.java:52]: recover this API

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/PenetrateAttachmentSelector.java [hunk 3]
```diff
@@ -29,6 +29,7 @@ public interface PenetrateAttachmentSelector {
      *
      * @return attachment pass to next hop
      */
-    Map<String, Object> select();
+    Map<String, Object> select(Invocation invocation);
+    Map<String, Object> selectReverse(Invocation invocation);
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/PenetrateAttachmentSelector.java:33]: Add `RpcContextAttachment serverAttachement / clientResponse` param

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcContext.java [hunk 4]
```diff
@@ -56,7 +56,27 @@ public class RpcContext {
...
+        @Override
+        public RpcContextAttachment setObjectAttachment(String key, Object value) {
+            if (value == null) {
+                attachments.remove(key);
+            } else {
+                RpcContext.getServerResponseContext().setAttachment(key, value);
+                attachments.put(key, value);
+            }
+            return this;
+        }
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcContext.java:69]: create a new class file for this

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcContext.java [hunk 5]
```diff
@@ -56,7 +56,27 @@ public class RpcContext {
...
+    };
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcContext.java:70]: override all methods in RpcContextAttachment, and proxy to serverResponse / clientResponse

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        @Override
+        public AsyncContext getAsyncContext() {
+            return RpcContext.getClientResponseContext().getAsyncContext();
+        }
+
+        @Override
+        public String getAttachment(String key) {
+            return RpcContext.getClientResponseContext().getAttachment(key);
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java:51]: Get from `serverResponse` first and then fall back to `clientResponse`

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        @Override
+        public Object getObjectAttachment(String key) {
+            return RpcContext.getClientResponseContext().getObjectAttachment(key);
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java:56]: Get from `serverResponse` first and then fall back to `clientResponse`

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        @Override
+        public RpcContextAttachment setAttachment(String key, Object value) {
+            return RpcContext.getServerResponseContext().setAttachment(key, value);
+        }
+
+        @Override
+        public RpcContextAttachment removeAttachment(String key) {
+            return RpcContext.getServerResponseContext().removeAttachment(key);
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java:71]: Remove in both `serverResponse` and `clientResponse`

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        @Override
+        public Map<String, String> getAttachments() {
+            return RpcContext.getServerResponseContext().getAttachments();
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java:76]: Create a new map, add all attachments from `clientResponse` and `serverResponse`. `serverResponse` should override `clientResponse`.

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        @Override
+        public Map<String, Object> getObjectAttachments() {
+            return RpcContext.getServerResponseContext().getObjectAttachments();
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java:81]: Create a new map, add all attachments from `clientResponse` and `serverResponse`. `serverResponse` should override `clientResponse`.

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        @Override
+        public RpcContextAttachment setObjectAttachments(Map<String, Object> attachment) {
+            return RpcContext.getServerResponseContext().setObjectAttachments(attachment);
+        }
+
+        @Override
+        public void clearAttachments() {
+            RpcContext.getServerResponseContext().clearAttachments();
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java:96]: Clear both `serverResponse` and `clientResponse`

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        @Override
+        public Map<String, Object> get() {
+            return RpcContext.getServerResponseContext().get();
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java:101]: Deprecated method. Proxy it to getAttachements()

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        @Override
+        public RpcContextAttachment set(String key, Object value) {
+            return RpcContext.getServerResponseContext().set(key, value);
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java:106]: Deprecated method. Proxy it to `setAttachment`

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        @Override
+        public RpcContextAttachment remove(String key) {
+            return RpcContext.getServerResponseContext().remove(key);
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java:111]: Deprecated method. Proxy it to `removeAttachment`

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        @Override
+        public Object get(String key) {
+            return RpcContext.getServerResponseContext().get(key);
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java:116]: Deprecated method. Proxy it to `getAttachment`

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        @Override
+        public RpcContextAttachment copyOf(boolean needCopy) {
+            return RpcContext.getServerResponseContext().copyOf(needCopy);
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java:121]: Internal method, should not be called. Throw exception directly. BTW, check the usage.

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        @Override
+        protected boolean isValid() {
+            return RpcContext.getServerResponseContext().isValid();
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java:126]: Internal method, should not be called. Throw exception directly. BTW, check the usage.

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/ContextFilter.java [hunk 18]
```diff
@@ -58,6 +62,13 @@
  */
 @Activate(group = PROVIDER, order = Integer.MIN_VALUE)
 public class ContextFilter implements Filter, Filter.Listener {
+    private Set<PenetrateAttachmentSelector> supportedSelectors;
+
+    public ContextFilter() {
+        ApplicationModel applicationModel = ApplicationModel.defaultModel();
+        ExtensionLoader<PenetrateAttachmentSelector> selectorExtensionLoader = applicationModel.getExtensionLoader(PenetrateAttachmentSelector.class);
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/ContextFilter.java:69]:
```suggestion
    public ContextFilter(ApplicationModel applicationModel) {
        ExtensionLoader<PenetrateAttachmentSelector> selectorExtensionLoader = applicationModel.getExtensionLoader(PenetrateAttachmentSelector.class);
```

File: dubbo-rpc/dubbo-rpc-api/src/test/java/org/apache/dubbo/rpc/support/PenetrateAttachmentSelectorMock.java [hunk 19]
```diff
@@ -16,17 +16,24 @@
...
+    public Map<String, Object> select(Invocation invocation, RpcContextAttachment clientAttachment, RpcContextAttachment serverAttachment) {
         Map<String, Object> objectAttachments = RpcContext.getServerAttachment().getObjectAttachments();
         objectAttachments.put("testKey", "testVal");
         return objectAttachments;
     }
+
+    @Override
+    public Map<String, Object> selectReverse(Invocation invocation, RpcContextAttachment clientResponseContext, RpcContextAttachment serverResponseContext) {
+        return null;
+    }
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/test/java/org/apache/dubbo/rpc/support/PenetrateAttachmentSelectorMock.java:38]: Add some test cases for this

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

**Review by AlbumenJ — COMMENTED** [review]
Please enhance the description in https://dubbo.apache.org/zh/docs3-v2/java-sdk/advanced-features-and-usage/service/attachment

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/filter/support/ConsumerContextFilter.java [hunk 20]
```diff
@@ -75,15 +75,14 @@ public Result invoke(Invoker<?> invoker, Invocation invocation) throws RpcExcept
 
             if (CollectionUtils.isNotEmpty(supportedSelectors)) {
                 for (PenetrateAttachmentSelector supportedSelector : supportedSelectors) {
-                    Map<String, Object> selected = supportedSelector.select();
+                    Map<String, Object> selected = supportedSelector.select(invocation, RpcContext.getClientAttachment(), RpcContext.getServerAttachment());
                     if (CollectionUtils.isNotEmptyMap(selected)) {
                         ((RpcInvocation) invocation).addObjectAttachments(selected);
                     }
                 }
```
Comment (tracywenjin) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/filter/support/ConsumerContextFilter.java:82]: 第三方

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java [hunk 21]
```diff
@@ -0,0 +1,253 @@
...
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(rpcServerContextAttachment);
+        }
+
+        public AttachmentMap(RpcServerContextAttachment rpcServerContextAttachment) {
+            Map<String, String> attachmentMap = this;
+            attachmentMap.putAll(getAllAttachmentMap());
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServerContextAttachment.java:165]: what is the purpose of these two lines

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/AbstractServerCallListener.java [hunk 22]
```diff
@@ -50,12 +50,14 @@ public void invoke() {
         RpcContext.restoreCancellationContext(cancellationContext);
         InetSocketAddress remoteAddress = (InetSocketAddress) invocation.getAttributes()
             .remove(AbstractServerCall.REMOTE_ADDRESS_KEY);
+        RpcContext.getServerResponseContext().setRemoteAddress(remoteAddress);
         RpcContext.getServerContext().setRemoteAddress(remoteAddress);
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/AbstractServerCallListener.java:54]:
```suggestion
        RpcContext.getServiceContext().setRemoteAddress(remoteAddress);
```

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/AbstractServerCallListener.java [hunk 23]
```diff
@@ -50,12 +50,14 @@ public void invoke() {
...
         String remoteApp = (String) invocation.getAttributes()
             .remove(TripleHeaderEnum.CONSUMER_APP_NAME_KEY);
         if (null != remoteApp) {
+            RpcContext.getServerResponseContext().setRemoteApplicationName(remoteApp);
             RpcContext.getServerContext().setRemoteApplicationName(remoteApp);
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/AbstractServerCallListener.java:59]:
```suggestion
        RpcContext.getServiceContext().setRemoteApplicationName(remoteApp);
```

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/AbstractServerCallListener.java [hunk 24]
```diff
@@ -50,12 +50,14 @@ public void invoke() {
...
-        }
+        };
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/AbstractServerCallListener.java:60]:
```suggestion
        }
```

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

**Review by AlbumenJ — APPROVED** [review]
LGTM

**Pull request merged**